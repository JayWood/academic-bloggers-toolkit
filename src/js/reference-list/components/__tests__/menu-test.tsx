import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { observable } from 'mobx';
import * as React from 'react';
import Menu, { dynamicOptionHeightHandler, renderer as Renderer } from '../menu';

const styles = {
    renamed: {},
    styles: [
        {
            label: 'American Medical Association',
            value: 'american-medical-association',
            id: 'american-medical-association',
        },
        { label: 'APA 5th Edition', value: 'apa-5th', id: 'apa-5th' },
    ],
};

const custom_csl = {
    CSL: '',
    label: 'Test Label',
    value: 'test-value',
};

const setupMenu = (isOpen: boolean = true) => {
    const spy = jest.fn();
    const component = mount(
        <Menu
            isOpen={observable(isOpen)}
            cslStyle={observable('american-medical-association')}
            itemsSelected={true}
            onSubmit={spy}
        />,
    );
    return {
        component,
        instance: component.instance() as Menu,
        importBtn: component.find('button#OPEN_IMPORT_DIALOG'),
        spy,
    };
};

const setupRenderer = (label: string, value = 'test', focusedOption = false) => {
    const option = { label, value };
    const focus = jest.fn();
    const select = jest.fn();
    const component = shallow(
        <Renderer
            style={{}}
            focusOption={focus}
            focusedOption={focusedOption === true ? option : {}}
            option={option}
            selectValue={select}
        />,
    );
    return {
        component,
        focus,
        select,
    };
};

describe('<Menu />', () => {
    const setup = setupMenu;
    beforeEach(() => {
        window.ABT = {
            ...window.ABT,
            styles,
            custom_csl,
        };
        jest.resetAllMocks();
    });
    it('should match snapshots', () => {
        let { component } = setup();
        expect(toJSON(component)).toMatchSnapshot();

        ({ component } = setup(false));
        expect(toJSON(component)).toMatchSnapshot();
    });
    it('should handle basic interactions', () => {
        const { component, importBtn, spy } = setup();
        expect(spy).toHaveBeenCalledTimes(0);
        importBtn.simulate('click');
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith({ kind: 'OPEN_IMPORT_DIALOG' });
        (component.instance() as any).handleSelect({
            label: 'APA 5th Edition',
            value: 'apa-5th',
            id: 'apa-5th',
        });
        expect(spy).toHaveBeenCalledTimes(2);
        expect(spy.mock.calls[1]).toEqual([{ data: 'apa-5th', kind: 'CHANGE_STYLE' }]);
    });
    it('should handle empty data', () => {
        const { component, importBtn, spy } = setup();
        expect(spy).toHaveBeenCalledTimes(0);
        importBtn.simulate('click');
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith({ kind: 'OPEN_IMPORT_DIALOG' });
        (component.instance() as any).handleSelect([]);
        expect(spy).toHaveBeenCalledTimes(1);
    });
    it('should work without custom CSL defined', () => {
        window.ABT.custom_csl = { value: undefined, label: '' };
        const { instance } = setup();
        expect(instance.styles).toEqual(styles.styles);
    });
    it('should have functioning static methods', () => {
        expect(Menu.willEnter()).toEqual({ height: 0, scale: 0 });
        expect(Menu.willLeave()).toEqual({
            height: { damping: 25, precision: 0.01, stiffness: 300, val: 0 },
            scale: { damping: 25, precision: 0.01, stiffness: 300, val: 0 },
        });
    });
});
describe('dynamicOptionHeightHandler()', () => {
    it('should render the appropriate heights', () => {
        expect(dynamicOptionHeightHandler({ option: { label: 'aaa', value: '', id: 'a' } })).toBe(
            30,
        );
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(35), value: '', id: 'a' } }),
        ).toBe(30);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(36), value: '', id: 'a' } }),
        ).toBe(40);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(65), value: '', id: 'a' } }),
        ).toBe(40);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(66), value: '', id: 'a' } }),
        ).toBe(50);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(80), value: '', id: 'a' } }),
        ).toBe(50);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(81), value: '', id: 'a' } }),
        ).toBe(60);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(90), value: '', id: 'a' } }),
        ).toBe(60);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(91), value: '', id: 'a' } }),
        ).toBe(70);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(110), value: '', id: 'a' } }),
        ).toBe(70);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(111), value: '', id: 'a' } }),
        ).toBe(90);
        expect(
            dynamicOptionHeightHandler({ option: { label: 'a'.repeat(250), value: '', id: 'a' } }),
        ).toBe(90);
    });
});
describe('Menu Renderer', () => {
    const setup = setupRenderer;
    it('should render options with the correct heights', () => {
        let { component } = setup('a');
        expect(component.props().style.height).toBe(30);
        ({ component } = setup('a'.repeat(35)));
        expect(component.props().style.height).toBe(30);
        ({ component } = setup('a'.repeat(36)));
        expect(component.props().style.height).toBe(40);
        ({ component } = setup('a'.repeat(65)));
        expect(component.props().style.height).toBe(40);
        ({ component } = setup('a'.repeat(66)));
        expect(component.props().style.height).toBe(50);
        ({ component } = setup('a'.repeat(80)));
        expect(component.props().style.height).toBe(50);
        ({ component } = setup('a'.repeat(81)));
        expect(component.props().style.height).toBe(60);
        ({ component } = setup('a'.repeat(90)));
        expect(component.props().style.height).toBe(60);
        ({ component } = setup('a'.repeat(91)));
        expect(component.props().style.height).toBe(70);
        ({ component } = setup('a'.repeat(110)));
        expect(component.props().style.height).toBe(70);
        ({ component } = setup('a'.repeat(111)));
        expect(component.props().style.height).toBe(90);
        ({ component } = setup('a'.repeat(250)));
        expect(component.props().style.height).toBe(90);
    });
    it('should render the header row correctly', () => {
        const { component } = setup('label', 'header');
        const style: React.CSSProperties = {
            alignItems: 'center',
            backgroundColor: '#eee',
            borderBottom: '1px solid #ddd',
            cursor: 'default',
            display: 'flex',
            fontWeight: 400,
            height: 30,
            padding: '0 5px',
        };
        expect(component.props().style).toEqual(style);
    });
    it('should render with darker background when focused', () => {
        const { component } = setup('test', 'test', true);
        expect(component.props().style.backgroundColor).toBe('#f5f5f5');
    });
    it('should call the select handler when clicked', () => {
        const { component, select } = setup('test');
        expect(select).toHaveBeenCalledTimes(0);
        component.simulate('click');
        expect(select).toHaveBeenCalledTimes(1);
        expect(select.mock.calls[0]).toEqual([{ label: 'test', value: 'test' }]);
    });
    it('should call the focus handler when focused', () => {
        const { component, focus } = setup('test label');
        expect(focus).toHaveBeenCalledTimes(0);
        component.simulate('mouseOver');
        expect(focus).toHaveBeenCalledTimes(1);
        expect(focus.mock.calls[0]).toEqual([{ label: 'test label', value: 'test' }]);
    });
});
