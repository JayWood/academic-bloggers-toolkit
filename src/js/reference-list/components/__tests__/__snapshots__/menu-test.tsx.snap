// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should match snapshots 1`] = `
<Menu
  cslStyle={"american-medical-association"}
  isOpen={true}
  itemsSelected={true}
  onSubmit={[Function]}
>
  <TransitionMotion
    didLeave={[Function]}
    styles={
      Array [
        Object {
          "key": "menu",
          "style": Object {
            "height": Object {
              "damping": 20,
              "precision": 0.01,
              "stiffness": 300,
              "val": 85,
            },
            "scale": Object {
              "damping": 20,
              "precision": 0.01,
              "stiffness": 300,
              "val": 1,
            },
          },
        },
      ]
    }
    willEnter={[Function]}
    willLeave={[Function]}
  >
    <div
      className="abt-reflist-menu"
      style={
        Object {
          "height": 85,
          "maxHeight": 85,
          "opacity": undefined,
          "transform": "scaleY(1)",
          "transformOrigin": "top",
        }
      }
    >
      <div
        className="abt-panel abt-panel_subpanel"
      >
        <PanelButton
          data-tooltip="Import references from file"
          id="IMPORT_RIS"
          onClick={[Function]}
        >
          <a
            aria-label="Import references from file"
            className="abt-btn abt-btn_flat abt-btn_icon"
            data-tooltip="Import references from file"
            id="IMPORT_RIS"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              className="dashicons dashicons-media-code"
            />
          </a>
        </PanelButton>
        <PanelButton
          data-tooltip="Refresh reference list"
          id="REFRESH_PROCESSOR"
          onClick={[Function]}
        >
          <a
            aria-label="Refresh reference list"
            className="abt-btn abt-btn_flat abt-btn_icon"
            data-tooltip="Refresh reference list"
            id="REFRESH_PROCESSOR"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              className="dashicons dashicons-update"
            />
          </a>
        </PanelButton>
        <PanelButton
          data-tooltip="Delete all references"
          id="DESTROY_PROCESSOR"
          onClick={[Function]}
        >
          <a
            aria-label="Delete all references"
            className="abt-btn abt-btn_flat abt-btn_icon"
            data-tooltip="Delete all references"
            id="DESTROY_PROCESSOR"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              className="dashicons dashicons-trash"
            />
          </a>
        </PanelButton>
        <PanelButton
          data-tooltip="Insert Static Publication List"
          disabled={false}
          id="INSERT_STATIC_BIBLIOGRAPHY"
          onClick={[Function]}
        >
          <a
            aria-label="Insert Static Publication List"
            className="abt-btn abt-btn_flat abt-btn_icon"
            data-tooltip="Insert Static Publication List"
            disabled={false}
            id="INSERT_STATIC_BIBLIOGRAPHY"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              className="dashicons dashicons-list-view"
            />
          </a>
        </PanelButton>
        <PanelButton
          data-tooltip="Usage instructions"
          href="https://github.com/dsifford/academic-bloggers-toolkit/wiki"
          target="_blank"
        >
          <a
            aria-label="Usage instructions"
            className="abt-btn abt-btn_flat abt-btn_icon"
            data-tooltip="Usage instructions"
            href="https://github.com/dsifford/academic-bloggers-toolkit/wiki"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            target="_blank"
          >
            <span
              className="dashicons dashicons-editor-help"
            />
          </a>
        </PanelButton>
      </div>
      <div
        style={
          Object {
            "padding": "0 5px",
          }
        }
      >
        <VirtualizedSelect
          async={false}
          backspaceRemoves={false}
          clearable={false}
          id="style-select"
          maxHeight={200}
          onChange={[Function]}
          optionHeight={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "label": "Custom Style",
                "value": "header",
              },
              Object {
                "label": "Test Label",
                "value": "test-value",
              },
              Object {
                "label": "Pre-defined Styles",
                "value": "header",
              },
              Object {
                "label": "American Medical Association",
                "value": "american-medical-association",
              },
              Object {
                "label": "APA 5th Edition",
                "value": "apa-5th",
              },
            ]
          }
          style={
            Object {
              "cursor": "pointer",
              "fontWeight": 300,
            }
          }
          value={
            Object {
              "label": "American Medical Association",
              "value": "american-medical-association",
            }
          }
        >
          <Select
            addLabelText="Add \\"{label}\\"?"
            arrowRenderer={[Function]}
            async={false}
            autosize={true}
            backspaceRemoves={false}
            backspaceToRemoveMessage="Press backspace to remove {label}"
            clearAllText="Clear all"
            clearRenderer={[Function]}
            clearValueText="Clear value"
            clearable={false}
            deleteRemoves={true}
            delimiter=","
            disabled={false}
            escapeClearsValue={true}
            filterOptions={[Function]}
            id="style-select"
            ignoreAccents={true}
            ignoreCase={true}
            inputProps={Object {}}
            isLoading={false}
            joinValues={false}
            labelKey="label"
            matchPos="any"
            matchProp="any"
            maxHeight={200}
            menuBuffer={0}
            menuRenderer={[Function]}
            menuStyle={
              Object {
                "overflow": "hidden",
              }
            }
            multi={false}
            noResultsText="No results found"
            onBlurResetsInput={true}
            onChange={[Function]}
            onCloseResetsInput={true}
            optionComponent={[Function]}
            optionHeight={[Function]}
            optionRenderer={[Function]}
            options={
              Array [
                Object {
                  "label": "Custom Style",
                  "value": "header",
                },
                Object {
                  "label": "Test Label",
                  "value": "test-value",
                },
                Object {
                  "label": "Pre-defined Styles",
                  "value": "header",
                },
                Object {
                  "label": "American Medical Association",
                  "value": "american-medical-association",
                },
                Object {
                  "label": "APA 5th Edition",
                  "value": "apa-5th",
                },
              ]
            }
            pageSize={5}
            placeholder="Select..."
            required={false}
            scrollMenuIntoView={true}
            searchable={true}
            simpleValue={false}
            style={
              Object {
                "cursor": "pointer",
                "fontWeight": 300,
              }
            }
            tabSelectsValue={true}
            value={
              Object {
                "label": "American Medical Association",
                "value": "american-medical-association",
              }
            }
            valueComponent={[Function]}
            valueKey="value"
          >
            <div
              className="Select Select--single is-searchable has-value"
            >
              <div
                className="Select-control"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "cursor": "pointer",
                    "fontWeight": 300,
                  }
                }
              >
                <span
                  className="Select-multi-value-wrapper"
                  id="react-select-2--value"
                >
                  <Value
                    disabled={false}
                    id="react-select-2--value-item"
                    instancePrefix="react-select-2-"
                    onClick={null}
                    value={
                      Object {
                        "label": "American Medical Association",
                        "value": "american-medical-association",
                      }
                    }
                  >
                    <div
                      className="Select-value"
                    >
                      <span
                        aria-selected="true"
                        className="Select-value-label"
                        id="react-select-2--value-item"
                        role="option"
                      >
                        American Medical Association
                      </span>
                    </div>
                  </Value>
                  <Component
                    aria-activedescendant="react-select-2--value"
                    aria-expanded="false"
                    aria-haspopup="false"
                    aria-owns=""
                    className="Select-input"
                    minWidth="5"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    role="combobox"
                    value=""
                  >
                    <div
                      className="Select-input"
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-activedescendant="react-select-2--value"
                        aria-expanded="false"
                        aria-haspopup="false"
                        aria-owns=""
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        required={false}
                        role="combobox"
                        style={
                          Object {
                            "boxSizing": "content-box",
                            "width": "5px",
                          }
                        }
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      />
                    </div>
                  </Component>
                </span>
                <span
                  className="Select-arrow-zone"
                  onMouseDown={[Function]}
                >
                  <span
                    className="Select-arrow"
                    onMouseDown={[Function]}
                  />
                </span>
              </div>
            </div>
          </Select>
        </VirtualizedSelect>
      </div>
    </div>
  </TransitionMotion>
</Menu>
`;

exports[`<Menu /> should match snapshots 2`] = `
<Menu
  cslStyle={"american-medical-association"}
  isOpen={false}
  itemsSelected={true}
  onSubmit={[Function]}
>
  <TransitionMotion
    didLeave={[Function]}
    styles={Array []}
    willEnter={[Function]}
    willLeave={[Function]}
  />
</Menu>
`;
